src/mon/commands/Command.h

```diff
@@ -0,0 +1,330 @@
+#include <set>
+#include <sstream>
+
+#include <boost/intrusive_ptr.hpp>
+
+#include "mon/Monitor.h"
+#include "mon/MonOpRequest.h"
+
+#include "include/types.h"
+#include "include/encoding.h"
+#include "include/ceph_assert.h"
+#include "include/Context.h"
+
+#include "common/cmdparse.h"
+#include "common/RefCountedObj.h"
+
+using namespace std;
+
+/*
+ * CommandCommon
+ *
+ * Define members and functions common to commands.
+ */
+template<typename Service>
+struct Common
+{
+
+  explicit Common(
+      Monitor *_mon,
+      Service *_svc,
+      CephContext *_cct) :
+    mon(_mon),
+    service(_svc),
+    cct(_cct)
+  { }
+
+  virtual ~Common() { }
+
+ private:
+  Common() = delete;
+  void wait_for_finished_proposal(MonOpRequestRef op, Context *ctx) {
+    service->wait_for_finished_proposal(op, ctx);
+  }
+
+ protected:
+  Monitor *mon;
+  Service *service;
+  CephContext *cct;
+
+  version_t get_last_committed() {
+    return service->get_last_committed();
+  }
+
+ public:
+  bool reply_with_data(MonOpRequestRef op,
+		       int retcode,
+		       stringstream &ss,
+		       bufferlist &retdata,
+		       version_t version) {
+    string retstr;
+    getline(ss, retstr);
+
+    bool ret = true;
+
+    if (retcode < 0 && retstr.length() == 0) {
+      retstr = cpp_strerror(retcode);
+      ret = false;
+    }
+    mon->reply_command(op, retcode, retstr, retdata, version);
+    return ret;
+  }
+
+  virtual bool reply(
+      MonOpRequestRef op,
+      int retcode,
+      stringstream &ss,
+      version_t version) {
+    bufferlist retdata;
+    return reply_with_data(op, retcode, ss, retdata, version);
+  }
+
+  bool update(MonOpRequestRef op,
+	      stringstream &ss,
+	      version_t version) {
+
+    string retstr;
+    getline(ss, retstr);
+
+    wait_for_finished_proposal(op,
+	new Monitor::C_Command(mon, op, 0, retstr, version));
+    return true;
+  }
+
+  bool wait_retry(MonOpRequestRef op) {
+    wait_for_finished_proposal(op,
+	new Monitor::C_RetryMessage(mon, op));
+    return true;
+  }
+};
+
+
+/* Command
+ *
+ * Define a command. It will be templated with
+ *  T	    = a paxos service
+ *  Stable  = the typename of the paxos service's stable map
+ *  Pending = the typename of the paxos service's pending map
+ *
+ * Depending on a paxos service's specificities, Stable and Pending may be
+ * the same.
+ *
+ * This class doesn't use these types for much more than defining the prepare
+ * and preprocess functions to be called by derived classes, but still
+ * requires them to be defined to properly abstract command definitions.
+ */
+
+typedef std::shared_ptr<Formatter> FormatterRef;
+
+template<typename Service, typename ...Args>
+struct Command : public Common<Service>
+{
+
+  CephContext *cct;
+
+  explicit Command(
+      Monitor *_mon,
+      Service *_service,
+      CephContext *_cct) :
+    Common<Service>(_mon, _service, _cct)
+  { 
+    cct = this->cct;
+  }
+
+  
+  virtual ~Command() { };
+
+  virtual bool preprocess(
```

Comment:
This is an important juncture for design discussion, because whether or not we provide a list of parameters as an interface,
or choose to do something else (like use a context structure) makes a difference in several areas that aren't necessarily
immediately obvious. This is one of the most central discussions from an "ergonomics" point of view, and I want to explain
why I deliberately avoided exactly this approach and encourage you to consider-- although it's definitely not perfect--
using a context struture. I'm not saying that one is better than the other-- I like things about both approaches-- but
I want to be sure there's discussion about the weighting of this choice.

First, let's consider the advantages of explicit parameters: the main ones, as I see it, is that it provides the easiest
migration path between the existing commands and any restructuring. This is a pretty valuable thing, because it eases the initial
conversion of code. It also has the virtual of providing a consistent and explicit set of parameters: by looking at the
member function's signature in the interface, it's clear what is available. It also is straightforward to implement,
which is not a bad thing.

Another thing-- and probably my favorite thing-- that I like about using explicit parameters relates to the above, but it's
that everything just looks like a normal function parameter. It's "natural". On the other hand, using a context structure--
at the cost of typing "ctx." (or whatever)-- let's you explicitly see anything and everything that's going to touch outside
state, which I also think is nice. I'm quite torn between these.

But let's consider first some "ergonomics". If I want to write a new command, I need to copy all of the parameter list
explicitly; small potatoes compared to actually writing the command. But, if the list of parameters ever needs to be changed,
all command signatures and the actual call site must be revised. This may or not be a concern, but with a context structure
this is avoided and can be done silently. Additionally, the names of the parameters can be "guaranteed" as they are always
the same inside the structure (and can be referenced in one place).

But, it is kinda nice to write:

 cmd_result my_cmd::preprocess(const state& ctx, string_view cmd) { ... }

...and get going. You do have to look up what's available in the state type.

A further implication of this is that there's an unfortunate "friend" relationship between the outer classes that these
commands were "born" in (i.e. they've always had access to private members, of say, Mon, so there are "hidden" parameters).
This may not be the end of the world-- it's what's already effectively the case now-- but if one of the goals it to try to
reduce indirect accidents, it may be important.

Throwing things into a context structure allows this to be measured.

The solution that I arrived at is admittedly not exactly pretty, and has the cost of requiring any state to
basically be available to all functions (although for most of this, it's already the case), whether actually consumed or
not (in practice didn't turn out to be as bad as I'd thought it might be).

It's important to note that the stated design goals of this project are different than my earlier attempt-- in
particular, my goals included that we make the required state for a given command explicit, reduce passing bare pointers,
reduce magical values, and provide for explicit success/failure (not necessarily in that order). It was also a goal to
see what parameters needed to be const or non-const, and centralize the naming conventions of them across all commands. If
those aren't important, this approach may be less useful.

As a final note, starting with this approach and then later /moving/ toward a context parameter would be easier than going
the other direction: so in a sense, this MAY create a bit more work now but it's always something that can be changed
pretty easily in the future.

Reply:
(Please let me be absolutely clear that I'm not saying this needs to be done and I'm definitely not proposing that there be two different PRs in play-- I'm just wanting to be sure these exchanges are discussed and well-understood.)